helix-core/src/main/java/org/apache/helix/manager/zk/ZkBaseDataAccessor.java

```diff
@@ -423,6 +423,16 @@ public boolean update(String path, DataUpdater<T> updater, int options) {
    * sync update
    */
   public AccessResult doUpdate(String path, DataUpdater<T> updater, int options) {
+    return doUpdate(path, updater, options, ZkClient.TTL_NOT_SET);
+  }
+
+  /**
+   * sync update with ttl
+   *
+   * ttl is only used when creating new znode, hence if znode is already created with a ttl, further
+   * update operations will not update the znode ttl even if ttl is provided in the options
+   */
+  public AccessResult doUpdate(String path, DataUpdater<T> updater, int options, long ttl) {
```

Comment:
I checked other fns (ex: doCreate and also the other invocation of doUpdate) and they are having public as the scope.

Reply:
Normally if the fuc is not used outside class, we would set as private. 
Also more importantly, it is a bit odd to allow user to set `ttl` but not actually update ttl in `update`.  If user provide a new `ttl` value and the node already exit, this function will return `RetCode.OK `without actually change anything.

Reply:
as I said earlier there are several instances of `doCreate`, `doSet`, `doUpdate` which are declared public. I followed the similar convention. I am not familiar with the helix code base, this is my first PR. If you can guide me what exact change to make here, happy to correct it.

Reply:
+1 on @xyuanlu.

I can understand TTL for create. But why we need TTL for update? The right operation is to let the user delete the node and recreate with a TTL.

I am not confortable with this feature. Before we have strong justification. I dont think we should let this in. 

@jacoblukose It would be good to figure out why you would like to support this.

Reply:
All other apis have clear end to end logic. Like I mentioned earlier, I don't think the api make sense when user provide TTL. 

I will let @junkaixue  chime in as he has more knowledge in Helix than I do.

Reply:
@junkaixue @xyuanlu thanks for the questions. The ask is internal, @xyuanlu you can read up in https://jira01.corp.linkedin.com:8443/browse/DEPEND-55695. On a high level, I have also mentioned the ask in PR description. We have a usecase, to use the ZkBucketDataAccessor class, however the same is not equipped to work the ttl func of zk. So this PR, basically add the ttl support to ZkBucketDataAccessor and other corresponding changes. 

Now to the q on why doUpdate need a ttl value to be passed in is because, `ZkBucketDataAccessor.compressedBucketWrite()` fn relies on `ZkBaseDataAccessor.doUpdate()` fn call. Now if the znode doesnt exist if we look at the internal logic of  `ZkBaseDataAccessor.doUpdate()` fn, it make a call to `ZkBaseDataAccessor.doCreate`, so here we need to pass in the ttl value. Hence I overloaded the existing `doUpdate` fn to pass in ttl. 

On the other q of access scope being public for the new overloaded `doUpdate` method, its only used by the `ZkBucketDataAccessor` class hence, I have changed to package-private. I cannot change to private, because then there is no access from `ZkBucketDataAccessor` to call `ZkBaseDataAccessor.doUpdate()`

Reply:
Please review and let me know if there are questions to clarify.

Reply:
@junkaixue @xyuanlu following up ^^

Reply:
I am OK as long as @junkaixue feels OK.   :)

Reply:
I am not buying this. As I said, when a node is created, we should not update its property and change it from a persistent node to TTL node. 

If you need to do this, then it should be the application logic to make that smoothly migrate from Non TTL to TTL. Also, since this is the API support Helix. I dont see any use cases in HELIX to leverage this.

Accessor is Helix internal thing. If you would like to leverage for company specific feature and not necessary help Helix, please fork the repo and implement it internally.

Reply:
@xyuanlu thanks for reponding and on the ack. It will be helpful if you can also callout alternative approach here to improve the craft if you thinks thats seen comprised. Happy to correct it. 

@junkaixue thanks for responding, would you be kind enough to highlight the code part where `I am not buying this. As I said, when a node is created, we should not update its property and change it from a persistent node to TTL node.` you have the mentioned concern. Also an alternative on how this need to handled would be great. 

I think there is some misunderstanding, the usecase is not change nodes created with persistent mode to migrate to use ttl mode. Idea is to just expand the scope of `ZkBucketDataAccessor` to leverage ttl func.

Reply:
To be honest I think more time is needed when design the ZK update API to keep consistency. I dont have an solid idea in mind how. But I think the fundamental rule is keep the behavior simple, consistent and not ambiguous.  

- The Update API has a TTL value, but the result is - the ZNode may or may not have the passed in TTL value, is not a simple and consistent behavior.  However, I don't have an idea of how to achieve this, I think this is the reason why we don't have update with TTL in the first place. More effort is needed when design this - if ever needed.

Reply:
thanks @xyuanlu the only reason why `doUpdate` need a ttl value to be passed in is because the current code is using a `create` fn inside the `doUpdate`. If there is no `create` call made within `doUpdate` then there is no need to pass in the ttl as well, reason being, `ttl` can only be enforced during creation of znode.